File: functional/queens.swift
#!/usr/bin/env swift

var size = 0
if C_ARGC > 1 {
    size = String.fromCString(C_ARGV[1])!.toInt()!
}

func is_safe(x: [Int], y: Int) -> Bool {
    return !contains(x, y) && !contains((0..<x.count).map{x.count-$0-abs(y-x[$0])}, 0)
}

func queen(n: Int) -> [[Int]] {
    if n <= 0 {
        return [[]]
    } else {
        return queen(n - 1).map{x in (1...size).map{$0}.filter{y in is_safe(x, y)}.map{y in x+[y]}}.reduce([], +)
    }
}

println(join("\n", queen(size).map{join(" ", $0.map{$0.description})}))

File: functional/queens.kt
fun main(args: Array<String>) {
    fun nqueens(size: Int): List<List<Int>> {
        fun queen(n: Int): List<List<Int>> {
            if (n <= 0) {
                return listOf(listOf())
            } else {
                return queen(n-1).flatMap{x->(1..size).filter{y->!x.contains(y) && !x.mapIndexed{i,e->Math.abs(y-e)-(x.size-i)}.contains(0)}.map{y->x+listOf(y)}}
            }
        }
        return queen(size)
    }

    println(nqueens(args[0].toInt()).map{x->x.joinToString(" ")}.joinToString("\n"))
}
